src/sensors/softfusion/drivers/bmi220.h

```diff
@@ -0,0 +1,419 @@
+        // firmware upload
+        i2c.writeReg(Regs::InitCtrl::reg, Regs::InitCtrl::valueStartInit);
+        for (uint16_t pos=0; pos<sizeof(bmi220_firmware);)
+        {
+            // tell the device current position
+
+            // this thing is little endian, but it requires address in bizzare form
+            // LSB register is only 4 bits, while MSB register is 8bits
+            // also value requested is in words (16bit) not in bytes (8bit)
+
+            const uint16_t pos_words = pos >> 1; // convert current position to words
+            const uint16_t position = (pos_words & 0x0F) | ((pos_words << 4) & 0xff00);
+            i2c.writeReg16(Regs::InitAddr, position);
+            // write actual payload chunk
+            const uint16_t burstWrite = std::min(sizeof(bmi220_firmware) - pos, I2CImpl::MaxTransactionLength);
+            i2c.writeBytes(Regs::InitData, burstWrite, const_cast<uint8_t*>(bmi220_firmware + pos));
+            pos += burstWrite;
+        }
+        i2c.writeReg(Regs::InitCtrl::reg, Regs::InitCtrl::valueEndInit);
+        delay(140);
+
+        // leave fifo_self_wakeup enabled
+        i2c.writeReg(Regs::PwrConf::reg, Regs::PwrConf::valueFifoSelfWakeup);
+        // check if IMU initialized correctly
+        if (!(i2c.readReg(Regs::InternalStatus::reg) & Regs::InternalStatus::initializedBit))
+        {
+            // firmware upload fail or sensor not initialized
+            return false;
+        }
+
+        // read zx factor used to reduce gyro cross-sensitivity error
+        const uint8_t zx_factor_reg = i2c.readReg(Regs::RaGyrCas);
+        const uint8_t sign_byte = (zx_factor_reg << 1) & 0x80;
+        zxFactor = static_cast<int8_t>(zx_factor_reg | sign_byte);
+        return true;
+    }
+
+    void setNormalConfig(MotionlessCalibrationData &gyroSensitivity)
+    {
+        i2c.writeReg(Regs::GyrConf::reg, Regs::GyrConf::value);
+        i2c.writeReg(Regs::GyrRange::reg, Regs::GyrRange::value);
+
+        i2c.writeReg(Regs::AccConf::reg, Regs::AccConf::value);
+        i2c.writeReg(Regs::AccRange::reg, Regs::AccRange::value);
+
+        if (gyroSensitivity.valid)
+        {
+            i2c.writeReg(Regs::Offset6::reg, Regs::Offset6::value);
+            i2c.writeBytes(Regs::GyrUserGain, 3, &gyroSensitivity.x);
+        }
+
+        i2c.writeReg(Regs::PwrCtrl::reg, Regs::PwrCtrl::valueGyrAccTempOn);
+        delay(100); // power up delay
+        i2c.writeReg(Regs::FifoConfig0::reg, Regs::FifoConfig0::value);
+        i2c.writeReg(Regs::FifoConfig1::reg, Regs::FifoConfig1::value);
+
+        delay(4);
+        i2c.writeReg(Regs::Cmd::reg, Regs::Cmd::valueFifoFlush);
+        delay(2);
+    }
+
+    bool initialize(MotionlessCalibrationData &gyroSensitivity)
+    {
+        if (!restartAndInit()) {
+            return false;
+        }
+
+        setNormalConfig(gyroSensitivity);
+
+        return true;
+    }
+
+    void motionlessCalibration(MotionlessCalibrationData &gyroSensitivity)
+    {
```

Comment:
did you check if advanced stuff like CRT really works (non-zero values returned, unique for each sensor?)

Reply:
No, the CRT function not work. It returned all-zero values.
I don't have any source code or datasheet related to BMI220. So I don't know how to modified the CRT function.
According to serial console, it seems like only the CRT won't work.

![image](https://github.com/user-attachments/assets/c53ede4c-49b8-4e62-9be5-69dc02d938a7)

![image](https://github.com/user-attachments/assets/b382409b-b97d-4c54-b3e2-1eef3e38b0b9)